particle-iot/device-os
# [CI] Migration to CircleCI

### Description

As title says. Updates CI scripts to run on CircleCI and adds circleci/config.yml. The builds are now performed in the release-style manner similar to Concourse with release scripts building platforms + device-os-test-runner building tests. The artifacts are collected on CircleCI, see https://app.circleci.com/jobs/github/particle-iot/device-os/611/artifacts. The scripts are somewhat crappy, but this PR has minimal changes from existing Codefresh/Concourse pipeline, we can always refactor things later.

_No custom Particle orbs were used as public org on CircleCI does not have access to those._

This PR also supersedes #2378 and adds HTML coverage report to artifacts.

The base image for buildpacks has also been updated to Ubuntu 20.04 and host (not ARM) GCC has been updated to GCC 11. See references (those repositories have also been migrated to CircleCI).

There are a couple of problems with how CircleCI executes workflows (e.g. unable to fail a job and still force-run another one), but this is what we have to deal with.

The only remaining thing is potentially adding support for building and pushing buildpacks.

#### This PR also fixes an issue with bootloader builds for Gen 3 platforms being broken on `develop`

### Steps to Test

1. See [this branch/PR build in CircleCI](https://app.circleci.com/pipelines/github/particle-iot/device-os/245/workflows/8cf2a293-aefd-4cb9-8ab5-85ecfb73e496)
2. See report in #device-os-bots
3. See [failed build of `ci/test-broken` branch](https://app.circleci.com/pipelines/github/particle-iot/device-os/244/workflows/99824140-88cd-4ba4-8cb8-44f9effdf064)
4. See report in #device-os-bots

### References

- https://github.com/particle-iot/buildpack-base/pull/8
- https://github.com/particle-iot/buildpack-hal/pull/14
- https://github.com/particle-iot/firmware-buildpack-builder/pull/23
- Closes #2378

---

### Completeness
